README.md
# Flow Chat for YouTube Live

![badge](https://github.com/fiahfy/youtube-live-chat-flow/workflows/Web%20Extension%20CI/badge.svg)

> Chrome Extension for Flow Chat Messages on YouTube Live.

## Features

- Flow messages over the video.
- Change color, size and speed for messages.
- Show author and avatar on messages.
- Show super chats and super stickers.
- Filter banned words.
- Move the chat input to bottom controls on the video.
- Add helper menu buttons on the chat list.

## Screenshots
![screenshot](.github/img/screenshot2.png)

## Installation

1. Download `archive.zip` from [releases page](https://github.com/fiahfy/youtube-live-chat-flow/releases) and unzip this file.
2. Open the Extension Management page by navigating to `chrome://extensions`.
3. Enable Developer Mode by clicking the toggle switch next to **Developer mode**.
4. Click the **LOAD UNPACKED** button and select the unpacked directory named `app`.

## Development

```bash
# install dependencies
yarn

# watch files changed and reload extension
yarn dev
```


Changes.txt
Disabled font outline
Removed border radius

src/store/settings.ts
import { Module, VuexModule, Mutation } from 'vuex-module-decorators'
import {
  AuthorType,
  EmojiStyle,
  HeightType,
  MessageType,
  Overflow,
  Settings,
  StackDirection,
  Style,
} from '~/models'

const initialState: Settings = {
  background: false,
  backgroundOpacity: 0.4,
  bottomChatInputEnabled: true,
  chatVisible: true,
  delayTime: 0,
  displayTime: 5,
  emojiStyle: 'image',
  growBottomChatInputEnabled: false,
  heightType: 'flexible',
  lineHeight: 64,
  lines: 12,
  maxDisplays: 0,
  maxLines: 0,
  maxWidth: 200,
  opacity: 0.8,
  overflow: 'hidden',
  stackDirection: 'top_to_bottom',
  styles: {
    guest: {
      avatar: false,
      color: '#ffffff',
      template: 'one-line-without-author',
    },
    member: {
      avatar: true,
      color: '#ccffcc',
      template: 'one-line-without-author',
    },
    moderator: {
      avatar: true,
      color: '#ccccff',
      template: 'two-line',
    },
    owner: {
      avatar: true,
      color: '#ffffcc',
      template: 'two-line',
    },
    you: {
      avatar: true,
      color: '#ffcccc',
      template: 'one-line-with-author',
    },
  },
  visibilities: {
    guest: true,
    member: true,
    moderator: true,
    owner: true,
    you: true,
    'super-chat': true,
    'super-sticker': true,
    membership: true,
  },
}

@Module({ name: 'settings' })
export default class SettingsModule extends VuexModule {
  background = initialState.background
  backgroundOpacity = initialState.backgroundOpacity
  bottomChatInputEnabled = true
  chatVisible = true
  delayTime = initialState.delayTime
  displayTime = initialState.displayTime
  emojiStyle = initialState.emojiStyle
  // extendedStyle = initialState.extendedStyle
  growBottomChatInputEnabled = false
  heightType = initialState.heightType
  lineHeight = initialState.lineHeight
  lines = initialState.lines
  maxDisplays = initialState.maxDisplays
  maxLines = initialState.maxLines
  maxWidth = initialState.maxWidth
  opacity = initialState.opacity
  overflow = initialState.overflow
  stackDirection = initialState.stackDirection
  styles = initialState.styles
  visibilities = initialState.visibilities

  @Mutation
  updateStyle({
    authorType,
    ...params
  }: { authorType: AuthorType } & Partial<Style>): void {
    this.styles = {
      ...this.styles,
      [authorType]: {
        ...this.styles[authorType],
        ...params,
      },
    }
  }
  @Mutation
  setVisibility({
    type,
    visibility,
  }: {
    type: AuthorType | MessageType
    visibility: boolean
  }): void {
    this.visibilities[type] = visibility
  }
  @Mutation
  setBackground({ background }: { background: boolean }): void {
    this.background = background
  }
  @Mutation
  setBackgroundOpacity({
    backgroundOpacity,
  }: {
    backgroundOpacity: number
  }): void {
    this.backgroundOpacity = backgroundOpacity
  }
  @Mutation
  setBottomChatInputEnabled({
    bottomChatInputEnabled,
  }: {
    bottomChatInputEnabled: boolean
  }): void {
    this.bottomChatInputEnabled = bottomChatInputEnabled
  }
  @Mutation
  setChatVisible({ chatVisible }: { chatVisible: boolean }): void {
    this.chatVisible = chatVisible
  }
  @Mutation
  setDelayTime({ delayTime }: { delayTime: number }): void {
    this.delayTime = delayTime
  }
  @Mutation
  setDisplayTime({ displayTime }: { displayTime: number }): void {
    this.displayTime = displayTime
  }
  @Mutation
  setEmojiStyle({ emojiStyle }: { emojiStyle: EmojiStyle }): void {
    this.emojiStyle = emojiStyle
  }
  // @Mutation
  // setExtendedStyle({ extendedStyle }: { extendedStyle: string }): void {
  //   this.extendedStyle = extendedStyle
  // }
  @Mutation
  setGrowBottomChatInputEnabled({
    growBottomChatInputEnabled,
  }: {
    growBottomChatInputEnabled: boolean
  }): void {
    this.growBottomChatInputEnabled = growBottomChatInputEnabled
  }
  @Mutation
  setHeightType({ heightType }: { heightType: HeightType }): void {
    this.heightType = heightType
  }
  @Mutation
  setLineHeight({ lineHeight }: { lineHeight: number }): void {
    this.lineHeight = lineHeight
  }
  @Mutation
  setLines({ lines }: { lines: number }): void {
    this.lines = lines
  }
  @Mutation
  setMaxDisplays({ maxDisplays }: { maxDisplays: number }): void {
    this.maxDisplays = maxDisplays
  }
  @Mutation
  setMaxLines({ maxLines }: { maxLines: number }): void {
    this.maxLines = maxLines
  }
  @Mutation
  setMaxWidth({ maxWidth }: { maxWidth: number }): void {
    this.maxWidth = maxWidth
  }
  @Mutation
  setOpacity({ opacity }: { opacity: number }): void {
    this.opacity = opacity
  }
  @Mutation
  setOverflow({ overflow }: { overflow: Overflow }): void {
    this.overflow = overflow
  }
  @Mutation
  setStackDirection({
    stackDirection,
  }: {
    stackDirection: StackDirection
  }): void {
    this.stackDirection = stackDirection
  }
  @Mutation
  resetState(): void {
    for (const [k, v] of Object.entries(initialState)) {
      ;(this as any)[k] = v // eslint-disable-line @typescript-eslint/no-explicit-any
    }
  }
}


src/utils/message-parser.ts
import { Message } from '~/models'
import { getImageSourceAsync } from '~/utils/dom-helper'

const getBackgroundColor = (el: HTMLElement): string => {
  return getComputedStyle(el).backgroundColor
}

const parseCommonElements = async (
  el: HTMLElement
): Promise<{
  message: string | undefined
  author: string | undefined
  authorType: string | undefined
  avatarUrl: string | undefined
}> => {
  const author = el.querySelector('#author-name')?.textContent ?? undefined
  const authorType = el.getAttribute('author-type') ?? undefined
  const avatorImage = el.querySelector('#img') as HTMLImageElement | null
  const avatarUrl =
    (avatorImage && (await getImageSourceAsync(avatorImage))) ?? undefined
  const message = el.querySelector('#message')?.textContent ?? undefined

  return { message, author, authorType, avatarUrl }
}

const parseTextMessage = async (
  el: HTMLElement
): Promise<{
  html: string | undefined
  messageType: string
  message: string | undefined
  author: string | undefined
  authorType: string | undefined
  avatarUrl: string | undefined
}> => {
  const params = await parseCommonElements(el)

  const html = el.querySelector('#message')?.innerHTML

  return {
    ...params,
    html,
    messageType: 'text-message',
  }
}

const parsePaidMessage = async (
  el: HTMLElement
): Promise<{
  html: string | undefined
  backgroundColor: string | undefined
  subText: string | undefined
  messageType: string
  message: string | undefined
  author: string | undefined
  authorType: string | undefined
  avatarUrl: string | undefined
}> => {
  const params = await parseCommonElements(el)

  const html = el.querySelector('#message')?.innerHTML
  const subText = el.querySelector('#purchase-amount')?.textContent ?? undefined
  const card = el.querySelector('#card > #header') as HTMLElement | null
  const backgroundColor = (card && getBackgroundColor(card)) ?? undefined

  return {
    ...params,
    html,
    backgroundColor,
    subText,
    messageType: 'paid-message',
  }
}

const parsePaidSticker = async (
  el: HTMLElement
): Promise<{
  stickerUrl: string | undefined
  backgroundColor: string | undefined
  subText: string
  messageType: string
  message: string | undefined
  author: string | undefined
  authorType: string | undefined
  avatarUrl: string | undefined
}> => {
  const params = await parseCommonElements(el)

  const subText = el.querySelector('#purchase-amount-chip')?.textContent ?? ''
  const card = el.querySelector('#card') as HTMLElement | null
  const backgroundColor = (card && getBackgroundColor(card)) ?? undefined
  const stickerImage = el.querySelector(
    '#sticker > #img'
  ) as HTMLImageElement | null
  const stickerUrl =
    (stickerImage && (await getImageSourceAsync(stickerImage))) ?? undefined

  return {
    ...params,
    stickerUrl,
    backgroundColor,
    subText,
    messageType: 'paid-sticker',
  }
}

const parseMembershipItem = async (
  el: HTMLElement
): Promise<{
  html: string | undefined
  backgroundColor: string | undefined
  messageType: string
  message: string | undefined
  author: string | undefined
  authorType: string | undefined
  avatarUrl: string | undefined
}> => {
  const params = await parseCommonElements(el)

  const detailText =
    el.querySelector('#header-subtext')?.textContent ?? undefined
  const header = el.querySelector('#card > #header') as HTMLElement | null
  const backgroundColor = (header && getBackgroundColor(header)) ?? undefined

  return {
    ...params,
    html: detailText,
    backgroundColor,
    messageType: 'membership-item',
  }
}

export const parse = async (el: HTMLElement): Promise<Message | undefined> => {
  const tagName = el.tagName.toLowerCase()
  switch (tagName) {
    case 'yt-live-chat-text-message-renderer':
      return await parseTextMessage(el)
    case 'yt-live-chat-paid-message-renderer':
      return await parsePaidMessage(el)
    case 'yt-live-chat-paid-sticker-renderer':
      return await parsePaidSticker(el)
    case 'yt-live-chat-membership-item-renderer':
      return await parseMembershipItem(el)
  }
}


src/utils/flow-controller.ts
import { semaphore } from '@fiahfy/semaphore'
import { Message, Settings } from '~/models'
import { querySelectorAsync, waitAllImagesLoaded } from '~/utils/dom-helper'
import MessageSettings from '~/utils/message-settings'
import { parse } from '~/utils/message-parser'
import { render } from '~/utils/message-renderer'

const sem = semaphore()

const ClassName = {
  filterActivated: 'ylcfr-active',
  filteredMessage: 'ylcfr-filtered-message',
  deletedMessage: 'ylcfr-deleted-message',
}

interface Timeline {
  willAppear: number
  didAppear: number
  willDisappear: number
  didDisappear: number
}

class Limiter {
  private limits: number
  private count = 0
  private expireTime = Date.now()

  constructor(limits: number) {
    this.limits = limits
  }

  isOver() {
    const now = Date.now()
    if (now > this.expireTime) {
      this.count = 0
      this.expireTime = now + 1000
    }
    return ++this.count > this.limits
  }
}

export default class FlowController {
  private _enabled = false
  private _following = false
  private _settings: Settings | undefined
  private timelines: Timeline[][] = []
  private observer: MutationObserver | undefined
  private followingTimer = -1
  private cleanupTimer = -1
  private limiter: Limiter | undefined

  get enabled(): boolean {
    return this._enabled
  }

  set enabled(value: boolean) {
    this._enabled = value
    if (!this._enabled) {
      this.clear()
    }
  }

  get following(): boolean {
    return this._following
  }

  set following(value: boolean) {
    this._following = value
    if (value) {
      const scrollToBottom = () => {
        const hovered = !!document.querySelector('#chat:hover')
        if (hovered) {
          return
        }
        const scroller = document.querySelector('#item-scroller')
        if (scroller) {
          scroller.scrollTop = scroller.scrollHeight
        }
      }
      scrollToBottom()
      this.followingTimer = window.setInterval(scrollToBottom, 1000)
    } else {
      clearInterval(this.followingTimer)
    }
  }

  get settings(): Settings | undefined {
    return this._settings
  }

  set settings(value: Settings | undefined) {
    this._settings = value
    this.limiter = new Limiter(value?.maxDisplays ?? 0)
  }

  private async proceed(element: HTMLElement) {
    if (!this._enabled || !this.settings) {
      return
    }

    const video = parent.document.querySelector(
      'ytd-watch-flexy video.html5-main-video'
    ) as HTMLVideoElement | null
    if (!video || video.paused) {
      return
    }

    const container = parent.document.querySelector(
      '.html5-video-container'
    ) as HTMLElement | null
    if (!container) {
      return
    }

    const [lines, height] = this.getLinesAndHeight(
      video.offsetHeight,
      this.settings
    )

    const deleted = await this.validateDeletedMessage(element)
    if (deleted) {
      return
    }

    const message = await parse(element)
    if (!message) {
      return
    }

    const infoIcon = element.querySelector('.ylcf-info-icon')
    infoIcon && infoIcon.remove()

    if (this.settings.maxDisplays > 0 && this.limiter?.isOver()) {
      return
    }

    const me = await this.createMessageElement(message, height, this.settings)
    if (!me) {
      return
    }

    me.style.display = 'none'
    container.appendChild(me)
    await waitAllImagesLoaded(me)

    sem.acquire(async () => {
      if (!this.settings || video.paused) {
        me.remove()
        return
      }

      me.style.display = 'flex'

      const messageRows = Math.ceil(me.offsetHeight / Math.ceil(height))
      const containerWidth = container.offsetWidth
      const timeline = this.createTimeline(me, containerWidth, this.settings)

      const index = this.getIndex(lines, messageRows, timeline)
      if (index + messageRows > lines && this.settings.overflow === 'hidden') {
        me.remove()
        return
      }
      this.pushTimeline(timeline, index, messageRows)

      const z = Math.floor(index / lines)
      const y = (index % lines) + (z % 2 > 0 ? 0.5 : 0)
      const opacity = this.settings.opacity ** (z + 1)
      const top =
        this.settings.stackDirection === 'bottom_to_top'
          ? video.offsetHeight - height * (y + messageRows + 0.1)
          : height * (y + 0.1)

      me.style.top = `${top}px`
      me.style.opacity = String(opacity)
      me.style.zIndex = String(z + 1 + 11) // 11 is set to z-index on div.webgl

      const animation = this.createAnimation(me, containerWidth, this.settings)
      animation.onfinish = () => {
        me.remove()
      }
      animation.play()
    })
  }

  private getLinesAndHeight(
    videoHeight: number,
    settings: Settings
  ): [number, number] {
    let lines, height
    if (settings.heightType === 'fixed') {
      height = settings.lineHeight
      lines = Math.floor((videoHeight - height * 0.2) / height)
    } else {
      lines = settings.lines
      height = videoHeight / (lines + 0.2)
    }
    lines = settings.maxLines > 0 ? Math.min(settings.maxLines, lines) : lines
    return [lines, height]
  }

  private async validateDeletedMessage(element: HTMLElement): Promise<boolean> {
    const active = document.documentElement.classList.contains(
      ClassName.filterActivated
    )
    if (!active) {
      return false
    }
    const deleted = await new Promise<boolean>((resolve) => {
      const expireTime = Date.now() + 1000
      const timer = window.setInterval(() => {
        const filtered = element.classList.contains(ClassName.filteredMessage)
        if (filtered || Date.now() > expireTime) {
          clearInterval(timer)
          const deleted = element.classList.contains(ClassName.deletedMessage)
          resolve(deleted)
        }
      }, 10)
    })
    return deleted
  }

  private async createMessageElement(
    message: Message,
    height: number,
    settings: Settings
  ): Promise<HTMLElement | null> {
    const ms = new MessageSettings(message, settings)
    if (!ms.template) {
      return null
    }

    const element = render(ms.template, {
      ...message,
      author: ms.author ? message.author : undefined,
      avatarUrl: ms.avatar ? message.avatarUrl : undefined,
      fontColor: ms.fontColor,
      backgroundColor: ms.backgroundColor,
      height,
      width: settings.maxWidth,
      emojiStyle: settings.emojiStyle,
    })

    if (!element) {
      return null
    }

    element.classList.add('ylcf-flow-message')

    return element
  }

  private createTimeline(
    element: HTMLElement,
    containerWidth: number,
    settings: Settings
  ): {
    willAppear: number
    didAppear: number
    willDisappear: number
    didDisappear: number
  } {
    const displayMillis = settings.displayTime * 1000
    const delayMillis = settings.delayTime * 1000
    const w = element.offsetWidth
    const v = (containerWidth + w) / displayMillis
    const t = w / v
    const n = Date.now()

    return {
      willAppear: n + delayMillis,
      didAppear: n + t + delayMillis,
      willDisappear: n + displayMillis - t + delayMillis,
      didDisappear: n + displayMillis + delayMillis,
    }
  }

  private createAnimation(
    element: HTMLElement,
    containerWidth: number,
    settings: Settings
  ): Animation {
    element.style.transform = `translate(${containerWidth}px, 0px)`

    const duration = settings.displayTime * 1000
    const delay = settings.delayTime * 1000
    const keyframes = [
      { transform: `translate(${containerWidth}px, 0px)` },
      { transform: `translate(-${element.offsetWidth}px, 0px)` },
    ]
    const animation = element.animate(keyframes, { duration, delay })
    animation.pause()
    return animation
  }

  private isDeniedIndex(index: number, lines: number): boolean {
    // e.g. if lines value is "12", denied index is "23", "47", "71" ...
    return index % (lines * 2) === lines * 2 - 1
  }

  private getIndex(
    lines: number,
    messageRows: number,
    timeline: Timeline
  ): number {
    let index = this.timelines.findIndex((_, i, timelines) => {
      const mod = (i + messageRows) % lines
      if (mod > 0 && mod < messageRows) {
        return false
      }
      return Array(messageRows)
        .fill(1)
        .every((_, j) => {
          if (this.isDeniedIndex(i + j, lines)) {
            return false
          }

          const ts = timelines[i + j]
          if (!ts) {
            return true
          }

          const t = ts[ts.length - 1]
          if (!t) {
            return true
          }

          return (
            t.didDisappear < timeline.willDisappear &&
            t.didAppear < timeline.willAppear
          )
        })
    })
    if (index === -1) {
      index = this.timelines.length
      const mod = (index + messageRows) % lines
      if (mod > 0 && mod < messageRows) {
        index += messageRows - mod
      }
      if (this.isDeniedIndex(index + messageRows - 1, lines)) {
        index += messageRows
      }
    }
    return index
  }

  private pushTimeline(
    timeline: Timeline,
    index: number,
    messageRows: number
  ): void {
    Array(messageRows)
      .fill(1)
      .forEach((_, j) => {
        const i = index + j
        if (!this.timelines[i]) {
          this.timelines[i] = []
        }
        this.timelines[i].push(timeline)
      })
  }

  async observe(): Promise<void> {
    this.observer?.disconnect()

    const items = await querySelectorAsync(
      '#items.yt-live-chat-item-list-renderer'
    )
    if (!items) {
      return
    }

    this.observer = new MutationObserver((mutations) => {
      mutations.forEach((mutation) => {
        const nodes = Array.from(mutation.addedNodes)
        nodes.forEach((node: Node) => {
          if (node instanceof HTMLElement) {
            this.proceed(node)
          }
        })
      })
    })
    this.observer.observe(items, { childList: true })

    this.cleanupTimer = window.setInterval(() => {
      this.timelines = this.timelines.map((timelines) => {
        return timelines.filter((timeline) => {
          return timeline.didDisappear > Date.now()
        })
      })
    }, 1000)
  }

  disconnect(): void {
    clearInterval(this.cleanupTimer)
    this.observer?.disconnect()
  }

  play(): void {
    parent.document.querySelectorAll('.ylcf-flow-message').forEach((e) => {
      e.getAnimations().forEach((a) => a.play())
    })
  }

  pause(): void {
    parent.document.querySelectorAll('.ylcf-flow-message').forEach((e) => {
      e.getAnimations().forEach((a) => a.pause())
    })
  }

  clear(): void {
    parent.document.querySelectorAll('.ylcf-flow-message').forEach((e) => {
      e.remove()
    })
    this.timelines = []
  }
}


src/utils/message-settings.ts
import Color from 'color'
import { AuthorType, Message, Settings, Style } from '~/models'
import { Template } from '~/utils/message-renderer'

export default class MessageSettings {
  private message: Message
  private settings: Settings

  constructor(message: Message, settings: Settings) {
    this.message = message
    this.settings = settings
  }

  private get yourName(): string {
    // if input control exists
    const span = document.querySelector('#input-container span#author-name')
    if (span?.textContent) {
      return span.textContent
    }
    // if input control is moved
    const movedSpan = parent.document.querySelector(
      '#input-container span#author-name'
    )
    if (movedSpan?.textContent) {
      return movedSpan.textContent
    }
    // otherwise
    const button = parent.document.querySelector(
      '.html5-video-player .ytp-chrome-top-buttons .ytp-watch-later-button'
    ) as HTMLElement | null
    // TODO: japanese only
    return (
      button?.getAttribute('title')?.replace(' として後で再生します', '') ?? ''
    )
  }

  private get authorType(): AuthorType {
    const author = this.message.author
    const you = author && author === this.yourName
    const authorType = you ? 'you' : this.message.authorType ?? 'guest'
    return (
      ['guest', 'member', 'moderator', 'owner', 'you'].includes(authorType)
        ? authorType
        : 'guest'
    ) as AuthorType
  }

  private get paid(): boolean {
    return ['paid-message', 'paid-sticker', 'membership-item'].includes(
      this.message.messageType ?? ''
    )
  }

  private get style(): Style {
    return this.settings.styles[this.authorType]
  }

  get template(): Template | undefined {
    switch (this.message.messageType) {
      case 'text-message':
        return this.settings.visibilities[this.authorType]
          ? this.style.template === 'two-line'
            ? 'two-line-message'
            : 'one-line-message'
          : undefined
      case 'paid-message':
        return this.settings.visibilities['super-chat']
          ? 'two-line-message'
          : undefined
      case 'paid-sticker':
        return this.settings.visibilities['super-sticker']
          ? 'sticker'
          : undefined
      case 'membership-item':
        return this.settings.visibilities['membership']
          ? 'two-line-message'
          : undefined
    }
  }

  get author(): boolean {
    switch (this.message.messageType) {
      case 'text-message':
        return this.style.template !== 'one-line-without-author'
      case 'paid-message':
        return true
      case 'paid-sticker':
        return true
      case 'membership-item':
        return true
      default:
        return false
    }
  }

  get avatar(): boolean {
    return this.paid ? true : this.style.avatar
  }

  get fontColor(): string {
    return this.paid ? '#ffffff' : this.style.color
  }

  get backgroundColor(): string | undefined {
    const backgroundColor = this.paid
      ? this.message.backgroundColor
      : this.settings.background
      ? 'black'
      : undefined

    if (!backgroundColor) {
      return undefined
    }

    try {
      const o = new Color(backgroundColor).object()
      const opacity = this.settings.backgroundOpacity
      return `rgba(${o.r}, ${o.g}, ${o.b}, ${opacity})`
    } catch (e) {
      // parse error by invalid background color
      return undefined
    }
  }
}


src/utils/message-renderer.ts
import Color from 'color'
import { EmojiStyle } from '~/models'

// e.g. https://yt3.ggpht.com/-TusVtXdhftI/AAAAAAAAAAI/AAAAAAAAAAA/OCgsPx8gmAk/s32-c-k-no-mo-rj-c0xffffff/photo.jpg
const resizeAvatarUrl = (url: string, size: number): string => {
  return url.replace(/(\/s)\d+([^/]+\/photo\.jpg)$/, `$1${Math.ceil(size)}$2`)
}

// e.g. https://yt3.ggpht.com/8ucCFoLkjQ9pVsfaKuFnHqIQqFjSr7Bht0dVuptyntxe-t6uej1BfH_vTk-cn1nsZdXjkjqipg=w24-h24-c-k-nd
const resizeEmojiUrl = (url: string, size: number): string => {
  return url.replace(
    /(=w)\d+(-h)\d+([^=]+)$/,
    `$1${Math.ceil(size)}$2${Math.ceil(size)}$3`
  )
}

// e.g. https://lh3.googleusercontent.com/kgcJnLI6rRPD1Jm7xko7FNnl0k9qVFGzNvu8TmtTcAs4vHwigbTfa0N7N98r1TfqUPfHfRRln47UiRbeCr3Z=s40-rp
const resizeStickerUrl = (url: string, size: number): string => {
  return url.replace(/(=s)\d+([^=]+)$/, `$1${Math.ceil(size)}$2`)
}

const renderAvatar = (url: string, height: number): HTMLImageElement => {
  const el = document.createElement('img')
  el.src = resizeAvatarUrl(url, height)
  el.style.height = `${height}px`
  el.style.borderRadius = '50%'
  el.style.objectFit = 'cover'
  return el
}

const renderAuthor = (author: string, height: number): HTMLSpanElement => {
  const textHeight = height * 0.8
  const padding = height * 0.1
  const el = document.createElement('span')
  el.style.fontSize = `${textHeight * 0.9}px`
  el.style.lineHeight = `${textHeight}px`
  el.style.paddingTop = `${padding}px`
  el.style.paddingBottom = `${padding}px`
  el.style.minWidth = '0'
  el.style.overflow = 'hidden'
  el.style.textOverflow = 'ellipsis'
  el.style.maxWidth = '100%'
  el.textContent = author
  return el
}

const renderMessage = (
  html: string,
  height: number,
  emojiStyle: EmojiStyle
): HTMLSpanElement | null => {
  const el = document.createElement('span')
  el.style.minWidth = '0'
  el.style.overflow = 'hidden'
  el.style.textOverflow = 'ellipsis'
  el.style.maxWidth = '100%'
  el.innerHTML = html

  switch (emojiStyle) {
    case 'image':
      Array.from(el.querySelectorAll('img')).forEach((e) => {
        e.src = resizeEmojiUrl(e.src, height)
        e.style.height = `${height}px`
        e.style.verticalAlign = 'bottom'
      })
      break
    case 'text':
      Array.from(el.querySelectorAll('img')).forEach((e) => {
        let alt = e.getAttribute('alt') ?? ''
        if (alt.length > 2) {
          alt = `:${alt}:`
        }
        const text = document.createTextNode(alt)
        e.parentElement?.replaceChild(text, e)
      })
      break
    case 'none':
      Array.from(el.querySelectorAll('img')).forEach((e) => {
        e.remove()
      })
      if (!el.textContent?.trim()) {
        return null
      }
      break
  }

  return el
}

const renderStickerImage = (url: string, height: number): HTMLImageElement => {
  const el = document.createElement('img')
  el.src = resizeStickerUrl(url, height)
  el.style.height = `${height}px`
  return el
}

const renderOneLineMessage = ({
  html,
  author,
  subText,
  avatarUrl,
  fontColor,
  backgroundColor,
  height,
  width,
  emojiStyle,
}: {
  html?: string
  author?: string
  subText?: string
  avatarUrl?: string
  fontColor?: string
  backgroundColor?: string
  height: number
  width: number
  emojiStyle: EmojiStyle
}): HTMLDivElement | null => {
  const el = document.createElement('div')
  el.style.color = fontColor ?? 'white'
  el.style.fontSize = `${height * 0.8 * 0.9}px`
  el.style.lineHeight = `${height * 0.8}px`
  el.style.padding = `${height * 0.1}px ${height * 0.2}px`
  el.style.backgroundColor = backgroundColor ?? 'transparent'
  el.style.maxWidth = width > 0 ? `${width}%` : 'unset'

  if (avatarUrl) {
    const avatar = renderAvatar(avatarUrl, height * 0.8)
    avatar.style.marginRight = `${height * 0.2}px`
    el.append(avatar)
  }

  if (author || subText) {
    const a = renderAuthor(
      (author ?? '') + (author && subText ? ' - ' : '') + (subText ?? ''),
      height * 0.8
    )
    el.append(a)

    const colon = document.createElement('span')
    colon.style.marginRight = `${height * 0.2}px`
    colon.textContent = ':'
    el.append(colon)
  }

  const message = renderMessage(html ?? '', height * 0.8, emojiStyle)
  if (message) {
    el.append(message)
  }

  if (!message && !subText) {
    return null
  }

  return el
}

const renderTwoLineMessage = ({
  html,
  author,
  subText,
  avatarUrl,
  fontColor,
  backgroundColor,
  height,
  width,
  emojiStyle,
}: {
  html?: string
  author?: string
  subText?: string
  avatarUrl?: string
  fontColor?: string
  backgroundColor?: string
  height: number
  width: number
  emojiStyle: EmojiStyle
}): HTMLDivElement | null => {
  const el = document.createElement('div')
  el.style.color = fontColor ?? 'white'
  el.style.fontSize = `${height * 0.8 * 0.9}px`
  el.style.lineHeight = `${height * 0.8}px`
  el.style.padding = `${height * 0.1}px ${height * 0.2}px`
  el.style.alignItems = 'start'
  el.style.backgroundColor = backgroundColor ?? 'transparent'
  el.style.maxWidth = width > 0 ? `${width}%` : 'unset'

  if (avatarUrl) {
    const avatar = renderAvatar(avatarUrl, height * 0.8)
    avatar.style.marginRight = `${height * 0.2}px`
    el.append(avatar)
  }

  const wrapper = document.createElement('div')
  wrapper.style.display = 'flex'
  wrapper.style.flexDirection = 'column'
  wrapper.style.alignItems = 'start'
  wrapper.style.minWidth = '0'
  wrapper.append(
    renderAuthor(
      (author ?? '') + (subText ? ` - ${subText}` : ''),
      height * 0.8
    )
  )

  let message
  if (html) {
    message = renderMessage(html ?? '', height * 0.8, emojiStyle)
    if (message) {
      message.style.marginTop = `${height * 0.1}px`
      wrapper.append(message)
    }
  }

  if (!message && !subText) {
    return null
  }

  el.append(wrapper)
  return el
}

const renderSticker = ({
  stickerUrl,
  author,
  subText,
  avatarUrl,
  fontColor,
  backgroundColor,
  height,
}: {
  stickerUrl?: string
  author?: string
  subText?: string
  avatarUrl?: string
  fontColor?: string
  backgroundColor?: string
  height: number
}): HTMLDivElement => {
  const el = document.createElement('div')
  el.style.color = fontColor ?? 'white'
  el.style.fontSize = `${height * 0.8 * 0.9}px`
  el.style.lineHeight = `${height * 0.8}px`
  el.style.padding = `${height * 0.1}px ${height * 0.2}px`
  el.style.alignItems = 'start'
  el.style.backgroundColor = backgroundColor ?? 'transparent'

  if (avatarUrl) {
    const avatar = renderAvatar(avatarUrl, height * 0.8)
    avatar.style.marginRight = `${height * 0.2}px`
    el.append(avatar)
  }

  const wrapper = document.createElement('div')
  wrapper.style.display = 'flex'
  wrapper.style.flexDirection = 'column'
  wrapper.style.alignItems = 'start'
  wrapper.append(renderAuthor(author ?? '', height * 0.8))
  const message = renderAuthor(subText ?? '', height * 0.8)
  message.style.marginTop = `${height * 0.1}px`
  wrapper.append(message)
  el.append(wrapper)

  if (stickerUrl) {
    const sticker = renderStickerImage(stickerUrl, height * 1.7)
    sticker.style.marginLeft = `${height * 0.2}px`
    el.append(sticker)
  }

  return el
}

export type Template = 'one-line-message' | 'two-line-message' | 'sticker'

type Params = {
  html?: string
  author?: string
  subText?: string
  avatarUrl?: string
  stickerUrl?: string
  fontColor?: string
  backgroundColor?: string
  height: number
  width: number
  emojiStyle: EmojiStyle
}

export const render = (
  template: Template,
  params: Params
): HTMLElement | null => {
  switch (template) {
    case 'one-line-message':
      return renderOneLineMessage(params)
    case 'two-line-message':
      return renderTwoLineMessage(params)
    case 'sticker':
      return renderSticker(params)
  }
}
